packages/grafana-ui/src/components/uPlot/Plot.tsx

```diff
@@ -1,83 +1,79 @@
 import isNumber from 'lodash/isNumber';
 import { UPlotConfigBuilder } from './config/UPlotConfigBuilder';
+import usePrevious from 'react-use/lib/usePrevious';
 
 // uPlot abstraction responsible for plot initialisation, setup and refresh
 // Receives a data frame that is x-axis aligned, as of https://github.com/leeoniya/uPlot/tree/master/docs#data-format
 // Exposes contexts for plugins registration and uPlot instance access
 export const UPlotChart: React.FC<PlotProps> = props => {
   const canvasRef = useRef<HTMLDivElement>(null);
-  const [plotInstance, setPlotInstance] = useState<uPlot>();
+  const plotInstance = useRef<uPlot>();
   const plotData = useRef<AlignedDataWithGapTest>();
-  const previousConfig = usePrevious(props.config);
-
-  // uPlot config API
-  const { currentConfig, registerPlugin } = usePlotConfig(props.width, props.height, props.timeZone, props.config);
-
-  const initializePlot = useCallback(() => {
-    if (!currentConfig || !plotData) {
-      return;
-    }
-    if (!canvasRef.current) {
-      throw new Error('Missing Canvas component as a child of the plot.');
-    }
-
-    pluginLog('UPlotChart: init uPlot', false, 'initialized with', plotData.current, currentConfig);
-    const instance = new uPlot(currentConfig, plotData.current, canvasRef.current);
-
-    setPlotInstance(instance);
-  }, [setPlotInstance, currentConfig]);
-
+  const prevProps = usePrevious(props);
+  const prevBuilder = usePrevious(props.config);
+  const { isConfigReady, currentConfig, registerPlugin } = usePlotConfig(
+    props.width,
+    props.height,
+    props.timeZone,
+    props.config
+  );
   const getPlotInstance = useCallback(() => {
-    if (!plotInstance) {
+    if (!plotInstance.current) {
       throw new Error("Plot hasn't initialised yet");
     }
 
-    return plotInstance;
-  }, [plotInstance]);
+    return plotInstance.current;
+  }, []);
 
+  // Effect responsible for uPlot updates/initialization logic. It's performed whenever component's props have changed
   useLayoutEffect(() => {
-    plotData.current = {
-      data: props.data.frame.fields.map(f => f.values.toArray()) as AlignedData,
-      isGap: props.data.isGap,
-    };
-
-    if (plotInstance && previousConfig === props.config) {
-      updateData(props.data.frame, props.config, plotInstance, plotData.current.data);
+    // 0. Exit early if the component is not ready to initialize uPlot
+    if (!currentConfig.current || !canvasRef.current || props.width === 0 || props.height === 0) {
+      return;
     }
-  }, [props.data, props.config]);
-
-  useLayoutEffect(() => {
-    initializePlot();
-  }, [currentConfig]);
 
-  useEffect(() => {
-    const currentInstance = plotInstance;
-    return () => {
-      currentInstance?.destroy();
-    };
-  }, [plotInstance]);
+    // 1. When config is ready and there is no uPlot instance, create new uPlot and return
+    if (isConfigReady && !plotInstance.current) {
+      plotData.current = prepareData(props.data);
```

Comment:
looking at the code it looks like we can remove plotData as useRef, it is only written two, only time I can see we actually use it is  on line 65, when config or timezone change, and we could easily just call prepareData there as that should be infrequent